### PR TITLE
Verse: exit on triple Enter

### LINKED ISSUE #52773: Verse: Unable to exit via enter
Initially noted here https://github.com/WordPress/gutenberg/discussions/50129#discussioncomment-6489248. When trying to exit the Verse block, the only way is to use the option to insert after/before or to use the Inserter to insert a new block. This is replicated using 6.3 RC1:

https://github.com/WordPress/gutenberg/assets/26996883/eafef770-187d-4baf-8df5-6c8200b3a050

To address this, solutions have been offered https://github.com/WordPress/gutenberg/pull/52783#discussion_r1273380089 and https://github.com/WordPress/gutenberg/pull/52783#issuecomment-1645277213. TLDR: a single return keeps you in the selected block, and a subsequent return exits it. This matches behavior with the List block.

### PR DESCRIPTION
**Pull request by ellatrix**
## What?

Fixes #52773.

## Why?

It's hard to exit verse.

## How?

When the at the end of a verse block, and the two previous characters are line breaks, break out of the block. In other words, exit on triple Enter.

🍒 on the 🥧: when merging two verses, or paragraph with verse, add two line breaks.

### CODE REVIEW

File: packages/block-editor/src/components/rich-text/use-enter.js [hunk 1]
```diff
@@ -103,6 +102,25 @@ export function useEnter( props ) {
 						onSplitMiddle,
 						multilineTag,
 					} );
+				} else if (
+					onSplitAtEnd &&
+					start === end &&
+					end === text.length
+				) {
+					onSplitAtEnd();
+				} else if (
```
Comment (mtias) [inline packages/block-editor/src/components/rich-text/use-enter.js]: Can we quickly note the "why" of this condition for future proofing?

**Comment by jasmussen** [comment]
Actually typing out a short poem to me (despite the typo) validates that you are indeed right, triple-enter feels valid to me:

verse

**Review by jasmussen — APPROVED** [review]
Would appreciate a sanity check by others before you merge, but this IMO feels valid.

**Comment by ellatrix** [comment]
I'll add the same setting to Code and Preformatted, since these blocks are nearly identical (or should be). I'll also add some e2e tests.

**Comment by ellatrix** [comment]
@jasmussen Your video makes me think we need a caption option for Verse, maybe also even for Preformatted and Code.

**Comment by jasmussen** [comment]
I don't know that we necessarily need that, is it wrapped in a `figure`? 

But just to clarify, I added that byline mostly for illustrating why triple click to break out of the verse feels valid, the poem isn't actually complete, and if I had made real content I would have probably added some visual stylings. That isn't to say I would necessarily be opposed to the caption button, rather that I didn't immediately feel urgency for one, the Paragraph felt fine to me 👌

**Comment by mtias** [comment]
I don't think we need captions on these :) They are a bit clunky

**Comment by ellatrix** [comment]
Not captions in the current form, they are clunky

**Comment by stokesman** [comment]
This seems fine to me. I appreciate the apparent discoverability of this and see that it was well considered via discussion in the linked issue. I'd almost rather we have a uniform way to exit any block, with say <kbd>alt</kbd>+<kbd>enter</kbd>. But that would lack discoverability unless we were to implement some sort of hint when it seems like someone might be trying to exit a block.

**Comment by ellatrix** [comment]
@Mamaduka Does this look ok to you?

**Comment by Mamaduka** [comment]
@ellatrix, can you rebase on top of the latest trunk?

**Comment by ellatrix** [comment]
Done

**Review by Mamaduka — APPROVED** [review]
Thank you, @ellatrix! The changes look good to me ✅

**Comment by allenhuffman** [comment]
Is there a way to break out of a preformatted block, other than at the end? Consider a block of reformatted text:

> `10 PRINT "MY CODE IS IN BASIC"`
> `20 PRINT "REALLY SIMPLE BASIC"`
> `30 GOTO 10`

And you want to break it up and add comments --

> `10 PRINT "MY CODE IS IN BASIC"`
> 
> This is the first line and it prints a message.
> 
> `20 PRINT "REALLY SIMPLE BASIC"`
> `30 GOTO 10`

I was resorting to editing the HTML to break up large blocks like this. Is it possible to escape out of a PRE, splitting it up in to two PREs or something?

**Comment by jasmussen** [comment]
Perhaps triple-enter in the middle of a verse should split it too?

**Comment by allenhuffman** [comment]
> Perhaps triple-enter in the middle of a verse should split it too?

That seems to be how you can break up a numbered/ordered list. Or perhaps there is a way to add this to the block system, allowing the "..." menu to have a new option to "Split" -- maybe where "Add Before" and "Add After" are.

**Comment by jasmussen** [comment]
@ellatrix what do you think about triple-enter to split a verse in two?

**Comment by mtias** [comment]
For preformatted, maybe, but I'd avoid for verse where it can be a useful expressive tool.

**Comment by allenhuffman** [comment]
Is there a way to use the context menu to say "split block" or something? That way ENTER can continue to function as it always has, but someone could break up a paragraph, preformatted, code, etc.?

**Comment by jasmussen** [comment]
Good point about the expressiveness; if you want multiple linebreaks in a poem that's laid out in a particular way (e.g. Alice in Wonderland), it won't work that three linebreaks breaks it apart. 

Perhaps we can make it so if you press the down arrow at the end of the verse, it moves the cursor down into the appender, e.g. creates a new paragraph?

**Comment by mtias** [comment]
I think it's ok to break down at the end of it, I don't think we should force a split in the middle of it. Also, an explicit "split" action might be good to support.

**Pull request merged**